iceberg: json translation

nvartolomei opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

### Features

* Support for JSON (via JSON Schema) translation to Iceberg.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
datalake: translate json

nvartolomei committed:
iceberg/tests: extract gmock matchers for reuse

nvartolomei committed:
iceberg/conversion: add time rfc 3339 parsing utility

Needed for JSON Schema conversion.

nvartolomei committed:
datalake/coordinator: extra logging

nvartolomei committed:
iceberg/conversion: consistent naming for field_map

Finish incomplete refactoring. Consistent naming.

nvartolomei committed:
pp/sr: accessor for json schema doc

Want to access it from iceberg code.

This is similar to the accessors we have for proto/avro.

nvartolomei committed:
iceberg: json schema value translation

Use JSON Schema IR together with the JSON StAX parser to convert a JSON
value to an Iceberg value.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5b989250-d8c1-4ddb-a69e-be0d3d7110e3-->
<!--TEST RESULTS-->
<!--00fb3047-6c20-47a8-a1a2-fbf7fbea9bd5-->
<!--TEST RESULTS-->
<!--46c89028-228c-486b-8687-27b3be4f5c05-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68271">build#68271</a></summary>


| test_class                        | test_method                | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------------|:---------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterConfigTest                 | test_valid_settings        |                                                                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68271#0197c706-f799-4365-a86a-4ac029b0357e | FLAKY         | 15/21    | upstream reliability is '88.81856540084388'. current run reliability is '71.42857142857143'. drift is 17.38999 and the allowed drift is set to 50. The test should PASS |
| PartitionForceReconfigurationTest | test_basic_reconfiguration | {"acks": -1, "controller_snapshots": false, "restart": true}                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68271#0197c706-f799-4365-a86a-4ac029b0357e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| RandomNodeOperationsTest          | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68271#0197c704-e522-4417-a6b1-dc0387912f85 | FLAKY         | 19/21    | upstream reliability is '99.19028340080972'. current run reliability is '90.47619047619048'. drift is 8.71409 and the allowed drift is set to 50. The test should PASS  |
| TopicDeleteCloudStorageTest       | drop_lifecycle_marker_test | {"cloud_storage_type": 2}                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68271#0197c706-f797-4005-90f3-efb9ef7f5dee | FLAKY         | 19/21    | upstream reliability is '97.01754385964912'. current run reliability is '90.47619047619048'. drift is 6.54135 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68337">build#68337</a></summary>


| test_class                   | test_method                               | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------|:------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves   | {"cleanup_policy": "compact,delete"}                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68337#0197cc28-3629-4c91-9e67-a95a9736c9e8 | FLAKY         | 20/21    | upstream reliability is '98.3451536643026'. current run reliability is '95.23809523809523'. drift is 3.10706 and the allowed drift is set to 50. The test should PASS  |
| NodePoolMigrationTest        | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage_fast_moves"}                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68337#0197cc28-3628-4724-bcbd-5c8a77e4399e | FLAKY         | 20/21    | upstream reliability is '99.22680412371135'. current run reliability is '95.23809523809523'. drift is 3.98871 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest     | test_node_operations                      | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68337#0197cc28-3624-4008-905c-372dc4fec10a | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             |
| RandomNodeOperationsTest     | test_node_operations                      | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68337#0197cc28-3626-479c-9749-92bc7b03fd13 | FLAKY         | 20/21    | upstream reliability is '99.63099630996311'. current run reliability is '95.23809523809523'. drift is 4.3929 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest     | test_node_operations                      | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": true}          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68337#0197cc28-3629-4c91-9e67-a95a9736c9e8 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68431">build#68431</a></summary>


| test_class                   | test_method                | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------------|:---------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage         | {"cleanup_policy": "compact,delete"}                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68431#0197d156-8428-4f4e-b748-6c298c3a9c45 | FLAKY         | 20/21    | upstream reliability is '96.15384615384616'. current run reliability is '95.23809523809523'. drift is 0.91575 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest     | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68431#0197d156-8426-429b-baba-f15c9bec8075 | FLAKY         | 18/21    | upstream reliability is '99.62264150943396'. current run reliability is '85.71428571428571'. drift is 13.90836 and the allowed drift is set to 50. The test should PASS |
| TopicDeleteCloudStorageTest  | drop_lifecycle_marker_test | {"cloud_storage_type": 1}                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68431#0197d158-f38a-4eca-93a5-c1d29f7fddec | FLAKY         | 20/21    | upstream reliability is '97.12643678160919'. current run reliability is '95.23809523809523'. drift is 1.88834 and the allowed drift is set to 50. The test should PASS  |
</details>
<!--/body-->

nvartolomei committed:
datalake: placeholder for json translation

nvartolomei committed:
datalake: storage for json schema ir

Unlike avro and protobuf the IR for JSON Schema is not stored/available
within schema registry so we need to add support for it ourselves.

nvartolomei committed:
rptest: iceberg unicode test

dotnwat commented on src/v/iceberg/conversion/values_json.cc:
```diff
@@ -0,0 +1,266 @@
+        if (it == field_map.end()) {
+            // Key not found in the mapping, skip it.
+            // todo throw if extra keys not allowed?
+            // skip the value
+            co_await p.next();
```
cray

dotnwat commented on src/v/iceberg/conversion/time_rfc3339.h:
```diff
@@ -0,0 +1,33 @@
+/// See: https://datatracker.ietf.org/doc/html/rfc3339#section-5.6
+namespace iceberg::conversion::time_rfc3339 {
+
+basic_value_outcome<iceberg::timestamptz_value>
+date_time_str_to_timestampz(const ss::sstring&);
+
+basic_value_outcome<iceberg::date_value> date_str_to_date(const ss::sstring&);
+
+basic_value_outcome<iceberg::time_value> time_str_to_time(const ss::sstring&);
```
since we're referring to an RFC from 2002, does any of the functionality within these interfaces make sense to pluck out into a utils/strings/time library rather than keeping them locked up in the iceberg subsystem? or it's all iceberg specific? if the later, i wonder if naming it as the rfc is confusing (as opposed to just `iceberg/conversion/time.h` (which internally happens to use that rfc standard)?

nvartolomei replied:
It is tied to iceberg value representation and it is also tied to rfc 3339 rather than generic time utilities so to me the naming and location is perfect.

dotnwat commented on src/v/iceberg/conversion/values_json.cc:
```diff
@@ -0,0 +1,266 @@
+                  buf.size_bytes()));
+            }
+
+            iobuf_parser iop(std::move(buf));
```
there might be a slight, tiny, possibly non-existent performance advantage here if you used `const_iobuf_parser`

dotnwat commented on src/v/iceberg/conversion/values_json.cc:
```diff
@@ -0,0 +1,266 @@
+          "Expected start of JSON array for list type");
+    }
+
+    while (co_await p.next()) {
+        if (p.token() == experimental::serde::json::token::end_array) {
+            co_return;
+        }
+
+        co_await decode_field(
+          p, lv.elements.emplace_back(), lt.element_field->type, field_map);
+    }
```
looks like a tight loop (similar in other places above i saw)? maybe it would benefit from using `ss::do_until` rather than the overhead of a coroutine? same in other places. could always benchmark later and see it pops up on a profile--but tight loops are more likely to be candidates. i think travis d had a p99 talk about this last year

nvartolomei replied:
Will follow-up with a benchmark mirroring existing protobuf and avro and experiment with it.

nvartolomei replied:
Added to todo.

nvartolomei committed:
iceberg: test truncate json inputs

andrwng commented on src/v/datalake/record_schema_resolver.cc:
```diff
@@ -98,9 +101,44 @@ checked<resolved_type, type_resolver::errc> translate_protobuf_schema(
+    try {
+        auto& doc = document(json_def());
+        auto fc = iceberg::conversion::json_schema::frontend{};
+        // todo figure out
+        // todo is this cached anywhere?
```
Are you wondering about whether we can cache the IR? I bet the schema_cache could be adapted to, given the parallel between its value type and storage_t (not for this PR though)

dotnwat replied:
@nvartolomei

andrwng reviewed (approved):
Nice work! Nothing major from me

dotnwat commented on src/v/iceberg/conversion/values_json.cc:
```diff
@@ -0,0 +1,279 @@
+        }
+
+        // Decode the value for this key.
+        if (!co_await p.next()) {
+            throw value_conversion_exception(
+              "Failed to read next JSON token after key");
+        }
+
+        co_await decode_field(
+          p,
+          sv.fields[it->second.field_pos],
```
what prevents `it` from becoming invalid across `co_await` scheduling point? is it that the field_map remains unchanged for its lifetime?
